llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp

```diff
@@ -694,8 +694,13 @@ static Instruction *foldCtpop(IntrinsicInst &II, InstCombinerImpl &IC) {
   // Add range attribute since known bits can't completely reflect what we know.
   if (BitWidth != 1 && !II.hasRetAttr(Attribute::Range) &&
       !II.getMetadata(LLVMContext::MD_range)) {
-    ConstantRange Range(APInt(BitWidth, Known.countMinPopulation()),
-                        APInt(BitWidth, Known.countMaxPopulation() + 1));
+    unsigned KnownMinPop = Known.countMinPopulation();
```

Comment:
There is a phase ordering problem. We may get a more precise result later. See my previous comment https://github.com/llvm/llvm-project/pull/100870#issuecomment-2254213148.

Reply:
did not think that far, forgets that the checks id done multiple times.
but then it fells like the draft that you have is better then this solution so we do not need to execute isKnownNonZero if not needed.

Reply:
added the test and when for a combination of our solutions that reduced number of calls to intersectWith also used full set range if old is missing as it is faster in intersectWith